src/eval.c

```diff
@@ -172,6 +172,49 @@ int luaRedisReplicateCommandsCommand(lua_State *lua) {
     return 1;
 }
 
+#if defined(USE_JEMALLOC)
+/* When lua uses jemalloc, pass in luaAlloc as a parameter of lua_newstate. */
+static void *luaAlloc(void *ud, void *ptr, size_t osize, size_t nsize) {
+    UNUSED(ud);
+    UNUSED(osize);
+    if (nsize == 0) {
+        zfree_with_flags(ptr, MALLOCX_ARENA(server.lua_arena) | MALLOCX_TCACHE_NONE);
+        return NULL;
+    } else {
+        return zrealloc_with_flags(ptr, nsize, MALLOCX_ARENA(server.lua_arena) | MALLOCX_TCACHE_NONE);
```

Comment:
@oranagra 

https://github.com/redis/redis/actions/runs/8336654976/job/22814152491#step:6:6798

@sundb found that MALLOCX_TCACHE_NONE needs to be added here so that the defrag in 32 bit will not fail

Reply:
it no TCACHE_NONE, if the lua VM request memory from tcache, which may being the memory just release by non-lua.
from the following malloc-stats, the memory requested by lua VM from the tcahce is actually counted as arena[0],
which results in higher fragmentation of arena[0].
```
# without MALLOCX_TCACHE_NONE
# arena lua
bins:           size ind    allocated      nmalloc (#/sec)      ndalloc (#/sec)    nrequests   (#/sec)  nshards      curregs     curslabs  nonfull_slabs regs pgs   util       nfills (#/sec)     nflushes (#/sec)       nslabs     nreslabs (#/sec)      n_lock_ops (#/sec)       n_waiting (#/sec)      n_spin_acq (#/sec)  n_owner_switch (#/sec)   total_wait_ns   (#/sec)     max_wait_ns  max_n_thds
                     ---
                  16   1      1959840       123659    7728         1169      73       227657     14228        1       122490          479              0  256   1  0.998         5270     329            0       0          479          274      17            5941     371               0       0               0       0               1       0               0         0               0           0
                  24   2         7104          500      31          204      12          626        39        1          296            1              0  512   3  0.578          203      12            0       0            1            0       0             349      21               0       0               0       0               1       0               0         0               0           0
                  32   3           64            4       0            2       0           10         0        1            2            1              0  128   1  0.015            4       0            0       0            1            0       0             147       9               0       0               0       0               1       0               0         0               0           0
                  40   4       991760        55629    3476        30835    1927        56182      3511        1        24794           61             15  512   5  0.793         1567      97            0       0           61          161      10            2090     130               0       0               0       0               1       0               0         0               0           0
                  48   5      2399232        49984    3124            0       0        99774      6235        1        49984          196              0  256   3  0.996         1067      66            0       0          196            0       0            1403      87               0       0               0       0               1       0               0         0               0           0
                  56   6       759528        47292    2955        33729    2108        48090      3005        1        13563           40             15  512   7  0.662         1598      99            0       0           40          106       6            2116     132               0       0               0       0               1       0               0         0               0           0
                  64   7      3199744        49996    3124            0       0        49937      3121        1        49996          782              0   64   1  0.998         2623     163            0       0          782            0       0            3545     221               0       0               0       0               1       0               0         0               0           0
                  80   8      3997520        49969    3123            0       0        49872      3117        1        49969          196              0  256   5  0.995         2040     127            0       0          196            0       0            2376     148               0       0               0       0               1       0               0         0               0           0
```

```
# with MALLOCX_TCACHE_NONE
# arena lua
bins:           size ind    allocated      nmalloc (#/sec)      ndalloc (#/sec)    nrequests   (#/sec)  nshards      curregs     curslabs  nonfull_slabs regs pgs   util       nfills (#/sec)     nflushes (#/sec)       nslabs     nreslabs (#/sec)      n_lock_ops (#/sec)       n_waiting (#/sec)      n_spin_acq (#/sec)  n_owner_switch (#/sec)   total_wait_ns   (#/sec)     max_wait_ns  max_n_thds
                   8   0            8            4       0            3       0            4         0        1            1            1              0  512   1  0.001            0       0            0       0            2            0       0             148       9               0       0               0       0               1       0               0         0               0           0
                  16   1      2366976       147951    9863           15       1       147951      9863        1       147936          578              0  256   1  0.999            0       0            0       0          581            0       0          148686    9912               0       0               0       0               1       0               0         0               0           0
                  24   2      1190448        49938    3329          336      22        49938      3329        1        49602           97              0  512   3  0.998            0       0            0       0           98            0       0           50511    3367               0       0               0       0               1       0               0         0               0           0
                  32   3         1984        98774    6584        98712    6580        98774      6584        1           62            1              0  128   1  0.484            0       0            0       0            2            0       0          197627   13175               0       0               0       0               1       0               0         0               0           0
                  40   4       515360        49380    3292        36496    2433        49380      3292        1        12884           26              0  512   5  0.967            0       0            0       0           54           92       6           86069    5737               0       0               0       0               1       0               0         0               0           0
                  48   5            0        49320    3288        49320    3288        49320      3288        1            0            0              0  256   3      1            0       0            0       0        49314            0       0          148093    9872               0       0               0       0               1       0               0         0               0           0
                  56   6       719824        49329    3288        36475    2431        49329      3288        1        12854           26              0  512   7  0.965            0       0            0       0           53           90       6           85996    5733               0       0               0       0               1       0               0         0               0           0
                  64   7      3156032        49318    3287            5       0        49318      3287        1        49313          771              0   64   1  0.999            0       0            0       0          772            0       0           50234    3348               0       0               0       0               1       0               0         0               0           0
                  80   8      3945040        49316    3287            3       0        49316      3287        1        49313          193              0  256   5  0.998            0       0            0       0          194            0       0           49652    3310               0       0               0       0               1       0               0         0               0           0
```

Reply:
> it no TCACHE_NONE, if the lua VM request memory from tcache, which may being the memory just release by non-lua.

ohh, right :facepalm: 

well, maybe we can create our own private tcache and change tcache when starting ending (pausing) a script.
but i suppose that at first we should keep TCACHE_NONE.
let's just have some check to see the effect on lua performance.

Reply:
TCACHE_NONE does have a noticeable impact on performance, using a private tcache makes more sense and is easy.
```c
--- a/src/eval.c
+++ b/src/eval.c
@@ -178,10 +178,10 @@ static void *luaAlloc(void *ud, void *ptr, size_t osize, size_t nsize) {
     UNUSED(ud);
     UNUSED(osize);
     if (nsize == 0) {
-        zfree_with_flags(ptr, MALLOCX_ARENA(server.lua_arena) | MALLOCX_TCACHE_NONE);
+        zfree_with_flags(ptr, MALLOCX_ARENA(server.lua_arena) | MALLOCX_TCACHE(server.lua_tcache));
         return NULL;
     } else {
-        return zrealloc_with_flags(ptr, nsize, MALLOCX_ARENA(server.lua_arena) | MALLOCX_TCACHE_NONE);
+        return zrealloc_with_flags(ptr, nsize, MALLOCX_ARENA(server.lua_arena) | MALLOCX_TCACHE(server.lua_tcache));
     }
```

Reply:
for the record, use `MALLOCX_TCACHE_NONE` with `./src/redis-benchmark -P 10 -n 10000000 -r 100000000 eval "return __rand_int__" 0`

The performance seems to be worse than unstable (glic): 110075.18 -> 100861.35

Reply:
you mean jemalloc without tcache is worse than libc malloc?
i think the script you used is too short / simple.
how does jemalloc with separate tcache when compared to both?

Reply:
yes, it seem is worse than libc malloc, the script i used is simple, i will try to do it again when the code is ready.

i did not test the separate tcache one, since it has some issue. one is the this one:
```
=== REDIS BUG REPORT START: Cut & paste starting from here ===
15445:M 20 Mar 2024 11:29:44.995 # Redis 255.255.255 crashed by signal: 11, si_code: 1
15445:M 20 Mar 2024 11:29:44.995 # Accessing address: 0x12ee289e0
15445:M 20 Mar 2024 11:29:44.995 # Crashed running the instruction at: 0x10e7aa8ad

------ STACK TRACE ------
EIP:
0   redis-server                        0x000000010e7aa8ad je_tcache_bin_flush_small + 701

Backtrace:
0   libsystem_platform.dylib            0x00007ff80985737d _sigtramp + 29
1   ???                                 0x0000001000000012 0x0 + 68719476754
2   redis-server                        0x000000010e77dab1 je_arena_ralloc + 1521
3   redis-server                        0x000000010e77322a do_rallocx + 746
4   redis-server                        0x000000010e878fca luaAlloc + 218
5   redis-server                        0x000000010e7595aa luaM_growaux_ + 170
6   redis-server                        0x000000010e7528cc luaK_code + 172
7   redis-server                        0x000000010e75165b luaK_storevar + 315
8   redis-server                        0x000000010e75dbf7 assignment + 615
9   redis-server                        0x000000010e75b01b chunk + 4059
10  redis-server                        0x000000010e75ba0a block + 58
11  redis-server                        0x000000010e75a444 chunk + 1028
12  redis-server                        0x000000010e75c535 body + 277
13  redis-server                        0x000000010e75aecf chunk + 3727
14  redis-server                        0x000000010e759ee2 luaY_parser + 114
15  redis-server                        0x000000010e755642 f_parser + 66
16  redis-server                        0x000000010e754315 luaD_rawrunprotected + 85
17  redis-server                        0x000000010e755455 luaD_pcall + 69
18  redis-server                        0x000000010e7555d5 luaD_protectedparser + 69
19  redis-server                        0x000000010e74fc61 lua_load + 65
20  redis-server                        0x000000010e763af0 luaL_loadbuffer + 32
21  redis-server                        0x000000010e879317 scriptingInit + 775
22  redis-server                        0x000000010e87bb67 scriptCommand + 1047
23  redis-server                        0x000000010e7c72ba call + 474
24  redis-server                        0x000000010e7c9416 processCommand + 3942
25  redis-server                        0x000000010e7e53f6 processInputBuffer + 2646
26  redis-server                        0x000000010e7e4266 readQueryFromClient + 2070
27  redis-server                        0x000000010e8ff7ef connSocketEventHandler + 191
28  redis-server                        0x000000010e7b2883 aeProcessEvents + 771
29  redis-server                        0x000000010e7d992d main + 24445
30  dyld                                0x00007ff80949e386 start + 1942
```

when a bio thread do the lua_close, and the main thread do the lua_open, it will crash. it look like we can do the lua_close in a different thread. it is not thread-safe. i asked sundb for help and he is working on it.

Reply:
yeah, of course. we can't use that tcache from the bio thread. (even if we could, it would be wrong), we need to make an exception and avoid explicitly specifying it for mallocx when we're not in the main thread.